### PR TITLE
Addition of PopChild and PopChildren commands

### LINKED ISSUE #11186: Add PopChild and PopChildren command
## What problem does this solve or what need does it fill?

Currently, to remove children from a parent, `RemoveChildren` must be used on the parent. Neither of these are ideal in cases where the children are ordered in a particular manner and only the element(s) at the end needs to be popped off for 2 reasons. 

1. Using `RemoveChildren` in this case would require a separate query to get the children at the end to remove.
2. `RemoveChildren` would require two iterations through all the children, which wouldn't be efficient if there's a lot of children

## What solution would you like?

The solution I think would be best would be to add a `PopChild` and `PopChildren` command. `PopChildren` could use `drain` on the inner `SmallVec` while `PopChild` just uses `pop`, which solves both issues with using `RemoveChildren` above because this would only cause an iteration through the child or children to be popped. 

Also, having a `PopChild` and `PopChildren` keeps the API more consistent because there's already `AddChild` and `PushChildren` that adds element(s) to the end of the children. (On a separate note, it might be better to rename `AddChild` to `PushChild` but this would likely have to be in a separate issue.)

## What alternative(s) have you considered?

Removing a child or children from the end can be done with `RemoveChildren` but isn't ideal because the children to remove would need to be retreived in a separate query and isn't very efficient when there's a lot of children.

### PR DESCRIPTION
**Pull request by TirushOne**
This pull request is a direct result of issue #11186

I am not entirley firmiliar with bevy structure of commands so I might have done something wrong but it fine as far as I can tell

Fixes #11186

### CODE REVIEW

File: crates/bevy_hierarchy/src/child_builder.rs [hunk 1]
```diff
@@ -215,6 +215,57 @@ impl Command for RemoveChildren {
     }
 }
 
+///Command that removes the last Child from an entity
```
Comment (alice-i-cecile) [inline crates/bevy_hierarchy/src/child_builder.rs:218]:
```suggestion
/// Command that removes the last [`Child`] from an entity
```

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
I'm not 100% sure this should exist: when would one want to use this command?

In terms of implementation:

1. The command structs should be directly constructable if they're pub: there's no benefit to hiding their fields.
2. Users will expect an API to use this as a method directly with Commands / EntityCommands. Check how other commands use an extension trait.
3. Because this command is centered around a single entity, should it be an `EntityCommand` instead?

**Comment by TirushOne** [comment]
> I'm not 100% sure this should exist: when would one want to use this command?
> 
> In terms of implementation:
> 
> 1. The command structs should be directly constructable if they're pub: there's no benefit to hiding their fields.
> 2. Users will expect an API to use this as a method directly with Commands / EntityCommands. Check how other commands use an extension trait.
> 3. Because this command is centered around a single entity, should it be an `EntityCommand` instead?

you're going to have to run that by me again.

I would create a contructor for the command, but I can't find where the other commands have their contructors so I don't have any pattern to follow. If you would point to where I can find these command contructors and "extension traits" I could make some changes. searching through the bevy code base is like look for a niddle in a hay stack and I find what you're talking about within the hierarchy.

as far as I am awhere the point of this command is just to make removing children from parents in certain scenaros easier. rust user will already be firmila with "pop" when working with vectors so it seems like a itomatic command to have.

**Review by doonv — COMMENTED** [review]
Not bad.
By the way I recommend adding "Fixes https://github.com/bevyengine/bevy/issues/11186" to the top of your pull request's message. That tells github to automatically close that issue if this pull request gets merged.

File: crates/bevy_hierarchy/src/child_builder.rs [hunk 2]
```diff
@@ -215,6 +215,57 @@ impl Command for RemoveChildren {
...
+
+impl Command for PopChild {
+    fn apply(self, world: &mut World) {
+        let last_child_id = {
+            let parent = world.entity_mut(self.parent);
+            if let Some(children) = parent.get::<Children>() {
+                children.last().copied()
+            } else {
+                None
+            }
```
[the remark this answers is not in the document]
Comment (TirushOne) [reply crates/bevy_hierarchy/src/child_builder.rs:231]: thanks for the help with the changes

[the remark this answers is not in the document]
Comment (TirushOne) [reply crates/bevy_hierarchy/src/child_builder.rs:231]: ok well guess I'll just wait for this pr to get merged and if something is stopping that or needs to be fixed please let me know

**Comment by TirushOne** [comment]
I still don't understand though, why are some of these commands pub but there feilds are not (and I can't find any contructors) but PopChild and PopChildren have public feilds? it is becasue all of those aother commands are for internal use and not public use, why have them be pub? and if they are for public use, how are they getting contructed without public feilds and without contructors?

**Comment by doonv** [comment]
It appears other commands are accessed inside the `BuildChildren` extension trait. There are functions there that create and call their respective command, like this:
```rs
fn clear_children(&mut self) -> &mut Self {
    let parent = self.id();
    self.commands().add(ClearChildren { parent });
    self
}
```
I have no idea why the commands are public while their fields are not though, might be a good idea to create another pull request to fix that.

**Comment by alice-i-cecile** [comment]
I can see the argument for consistency (and I don't want to be discouraging), but I don't think this should be merged. 

Removing the last child is not an operation that users ever want to do: the operations that you want are "remove this specific entity" and "clear all other entities".

**Comment by TirushOne** [comment]
> I can see the argument for consistency (and I don't want to be discouraging), but I don't think this should be merged.
> 
> Removing the last child is not an operation that users ever want to do: the operations that you want are "remove this specific entity" and "clear all other entities".

I mean I don't know if this is convincing, but I guess becasue of the funtional nature of rust, and therefore bevy, parsing commands to the ecs requires a decent bit of boiler plate, not that much, but enough that I think it's worth a small amount of back end bloat with abtractions like "pop children" that removes the need for front end boiler plate like getting the children first and then removing them. It is up to you guys though. I will say if we are closing this pull then probobley also best to go and close the issues as well if it's not acctually an issue.

**Pull request closed without merging**